Add booking system tutorial (#1001)
encoredev/encore: docs/tutorials/booking-system.md

	fromDate, err := time.Parse("2006-01-02", from)
	if err != nil {
		return nil, err
	}

	const numDays = 7

	var slots []BookableSlot
	for i := 0; i < numDays; i++ {
		date := fromDate.AddDate(0, 0, i)
		daySlots, err := bookableSlotsForDay(date)
		if err != nil {
			return nil, err
		}
		slots = append(slots, daySlots...)
	}

	return &SlotsResponse{Slots: slots}, nil
}

func bookableSlotsForDay(date time.Time) ([]BookableSlot, error) {
	// 09:00
	availStartTime := pgtype.Time{
		Valid:        true,
		Microseconds: int64(9*3600) * 1e6,
	}
	// 17:00
	availEndTime := pgtype.Time{
		Valid:        true,
		Microseconds: int64(17*3600) * 1e6,
	}

	availStart := date.Add(time.Duration(availStartTime.Microseconds) * time.Microsecond)
	availEnd := date.Add(time.Duration(availEndTime.Microseconds) * time.Microsecond)

	// Compute the bookable slots in this day, based on availability.
	var slots []BookableSlot
	start := availStart
	for {
		end := start.Add(DefaultBookingDuration)
		if end.After(availEnd) {
			break
		}
		slots = append(slots, BookableSlot{
			Start: start,
			End:   end,
		})
		start = end
	}

	return slots, nil
}
```

> The availability is right now hardcoded to be 09:00 - 17:00 for each day. Later we'll add the functionality to set it for each day of the week.
> We are also returning slots that has already passed. Don't worry, we will come back and fix that as well.  

Reviewer: ```suggestion
The availability is currently hardcoded to be 09:00 - 17:00 for each day. Later we'll add the functionality to set it for each day of the week.
We are also returning time slots that have already passed. Don't worry, we'll come back and fix it later on.
```

> 🥐 Let's try it! Open up the Encore Development Dashboard running at <http://localhost:9400> and try calling

Reviewer: ```suggestion
🥐 Let's try it! Open up the Local Development Dashboard running at <http://localhost:9400> and try calling
```

the `booking.GetBookableSlots` endpoint, passing in `2024-12-01`.

If you prefer to use the terminal instead run `curl http://localhost:4000/slots/2024-12-01` in
a new terminal instead. Either way you should see the response:

```json
{
  "Slots": [
    {
      "start": "2024-12-01T09:00:00Z",
      "end": "2024-12-01T10:00:00Z"
    },
    {
      "start": "2024-12-01T10:00:00Z",
      "end": "2024-12-01T11:00:00Z"
    },
    {
      "start": "2024-12-01T11:00:00Z",
      "end": "2024-12-01T12:00:00Z"
    },
    ...
  ]
}
```

## 3. Book an appointment

> Next, we want to make it possible to book an appointment. To be able to do that we will need a database to store the bookings in. Encore makes it really simple to [create and use databases](/docs/primitives/databases) (both for local and cloud environments) but for this example we will also make use of [sqlc](https://sqlc.dev/) that will compile our SQL queries into type-safe Go code that we can use in our application.

Reviewer: ```suggestion
Next, we want to make it possible to book an appointment. We'll need a database to store the bookings in. Encore makes it really simple to [create and use databases](/docs/primitives/databases) (both for local and cloud environments), but for this example we will also make use of [sqlc](https://sqlc.dev/) that will compile our SQL queries into type-safe Go code that we can use in our application.
```

> 🥐 Let's create an SQL database for our booking service and the needed sqlc scaffolding. Create the following file structure:

Reviewer: ```suggestion
🥐 Let's create a SQL database for our booking service and the required sqlc scaffolding. Create the following file structure:
```

```
/my-app
└── booking                              // booking service (a Go package)
    ├── db                               // (New) db related files (directory)
    │   ├── migrations                   // (New) db migrations (directory)
    │   │   └── 1_create_table.up.sql    // (New) db migration schema
    │   └── query.sql                    // (New) SQL queries
    ├── sqlc.yaml                        // (New) sqlc config file
    ├── slots.go                         // booking service code
    └── helpers.go                       // booking service code
```

> 🥐  Naming of the database migration file is important, it must look something like: `1_<name>.up.sql`.

Reviewer: ```suggestion
🥐 Naming of the database migration file is important, it must look something like: `1_<name>.up.sql`.
```

    end_time TIMESTAMP NOT NULL,
    email TEXT NOT NULL,
    created_at TIMESTAMP NOT NULL DEFAULT NOW()
);
```

🥐 Next, install the sqlc library:

```shell
$ go get github.com/sqlc-dev/sqlc/cmd/sqlc
```

🥐 Next, we need to configure sqlc. Add the following contents to `sqlc.yaml`:

```yaml
version: "2"
sql:
  - engine: "postgresql"
    queries: "db/query.sql"
    schema: "./db/migrations"
    gen:
      go:
        package: "db"
        out: "db"
        sql_package: "pgx/v5"
```

This instructs sqlc to generate Go code from the queries in `db/query.sql` and models from the schemas in the `db/migrations` folder.

🥐 Let's create our first SQL queries. Add the following contents to `db/query.sql`:

```sql
-- name: InsertBooking :one
INSERT INTO booking (start_time, end_time, email)
VALUES ($1, $2, $3)
RETURNING *;

-- name: ListBookingsBetween :many
SELECT * FROM booking
WHERE start_time >= $1 AND end_time <= $2;

-- name: ListBookings :many
SELECT * FROM booking;

-- name: DeleteBooking :exec
DELETE FROM booking WHERE id = $1;

```

🥐 It's time for sqlc to shine! Run the following command in your terminal:

```shell
$ cd booking
$ sqlc generate
```

> Three files should now have been generated inside the `db` folder: `query.sql.go`, `db.go` & `models.go`. These files contain generated Go code and should not be manually edited. We will be adding more queries to `db/query.sql` later and then re-run `sqlc generate` to update the generated Go code.

Reviewer: ```suggestion
Three files should now have been generated inside the `db` folder: `query.sql.go`, `db.go` and `models.go`. These files contain generated Go code and should not be manually edited. We will be adding more queries to `db/query.sql` later and then re-run `sqlc generate` to update the generated Go code.
```

		StartTime: pgtype.Timestamp{Time: p.Start, Valid: true},
		EndTime:   pgtype.Timestamp{Time: p.Start.Add(DefaultBookingDuration), Valid: true},
		Email:     p.Email,
	})
	if err != nil {
		return eb.Cause(err).Code(errs.Unavailable).Msg("failed to insert booking").Err()
	}

	if err := tx.Commit(ctx); err != nil {
		return eb.Cause(err).Code(errs.Unavailable).Msg("failed to commit transaction").Err()
	}
	return nil
}
```

We are now using the generated type-safe `query.InsertBooking` function to make the database operation.

Notice the `encore:"sensitive"` tag on the `Email` field. This tells Encore to scrub this field so that the data is not viewable in the traces for deployed environments. This is useful for fields that contain [sensitive data](/docs/develop/api-schemas#sensitive-data) such as email addresses, passwords, etc.

🥐 Restart `encore run` to cause the database to be created, and then call the `booking.Book` endpoint:

```shell
$ curl -X POST 'http://localhost:4000/booking' -d '{"start": "2024-12-11T09:00:00Z", "email": "test@example.com"}'
```

Congratulations, you have now booked your first appointment!

## 4. Authentication

> In order to have an admin dashboard we need to add authentication to our application so that we can have protected endpoints. 

Reviewer: ```suggestion
To provide an admin dashboard for our booking system, we need to add authentication to our application so that we can have protected endpoints.

Keep in mind, in this tutorial we'll only include a very basic implementation.
```

```go
-- user/auth.go --
package user

import (
	"context"
	"encore.dev/beta/auth"
	"encore.dev/beta/errs"
)

type Data struct {
	Email string
}

type AuthParams struct {
	Authorization string `header:"Authorization"`
}

//encore:authhandler
func AuthHandler(ctx context.Context, p *AuthParams) (auth.UID, *Data, error) {
	if p.Authorization != "" {
		return "test", &Data{}, nil
	}
	return "", nil, errs.B().Code(errs.Unauthenticated).Msg("no auth header").Err()
}

```

This function is our [auth handler](/docs/develop/auth#the-auth-handler). An Encore applications can designate a special function to handle authentication, 
by defining a function and annotating it with `//encore:authhandler`. This annotation tells Encore to run the function whenever an 
incoming API call contains authentication data.

The auth handler is responsible for validating the incoming authentication data and returning an `auth.UID` (a string type representing a user id). 
The `auth.UID` can be whatever you wish, but in practice it usually maps directly to the primary key stored in a user table (either defined in the Encore service or in an external service like Firebase or Okta).

> In order to keep this example simple we will just approve any request containing a token that is not empty .

Reviewer: ```suggestion
In order to keep this example simple, we'll just approve any request containing a token that is not empty.
```

Next we will implement some of our auth endpoints and make use of our newly created auth handler.

## 5. Setting availability

> The availability is hardcoded to 9:00 - 17:00 at the moment. Let's change that by adding the functionality to make it possible set it dynamically. 

Reviewer: ```suggestion
Right now the availability is hardcoded to 9:00 - 17:00. Let's add the functionality to let our admin users customize this. 
```

> To do that we start by add another migration file, this time to add our `avilability` table.

Reviewer: ```suggestion
Let's start by adding another migration file, this time to create an `avilability` table.
```

		}

		start, err1 := strToTime(a.Start)
		end, err2 := strToTime(a.End)
		if err := errors.Join(err1, err2); err != nil {
			return eb.Cause(err).Code(errs.InvalidArgument).Msg("invalid start/end time").Err()
		} else if start.Valid != end.Valid {
			return eb.Code(errs.InvalidArgument).Msg("both start/stop must be set, or both null").Err()
		} else if start.Valid && start.Microseconds > end.Microseconds {
			return eb.Code(errs.InvalidArgument).Msg("start must be before end").Err()
		}

		err = qry.UpdateAvailability(ctx, db.UpdateAvailabilityParams{
			Weekday:   int16(weekday),
			StartTime: start,
			EndTime:   end,
		})
		if err != nil {
			return eb.Cause(err).Code(errs.Unavailable).Msg("failed to update availability").Err()
		}
	}

	err = tx.Commit(ctx)
	return errs.WrapCode(err, errs.Unavailable, "failed to commit transaction")
}
```

This file contains two endpoints, a setter and a getter. The `SetAvailability` endpoint is protected by the `auth` middleware which means that the user must be authenticated in order to call it. The `GetAvailability` endpoint is public and can be called without authentication.

> 🥐 Let's set the availability for each day of the week. Go to the development dashboard at <http://localhost:9400> and select the `booking.SetAvailability` endpoint in the API Explorer. For the request body, paste the following:

Reviewer: ```suggestion
🥐 Let's set the availability for each day of the week. Open the Development Dashboard at <http://localhost:9400> and select the `booking.SetAvailability` endpoint in the API Explorer. For the request body, paste the following:
```

```json
{
    "Availability": [{
        "start": "09:30",
        "end": "17:00"
    },{
        "start": "09:00",
        "end": "17:00"
    },{
        "start": "09:00",
        "end": "18:00"
    },{
        "start": "08:30",
        "end": "18:00"
    },{
        "start": "09:00",
        "end": "17:00"
    },{
        "start": "09:00",
        "end": "17:00"
    },{
        "start": "09:30",
        "end": "16:30"
    }]
}
```

<Callout type="info">

> Remember to not leave the auth token empty because then the auth handler will then reject the request. You can use any value for the auth token.

Reviewer: ```suggestion
Don't leave the auth token empty, it will cause the auth handler to reject the request. You can use any value for the auth token.
```

</Callout>

> Now try retrieving the availability by calling the `booking.GetAvailability` endpoint through the API Explorer.

Reviewer: ```suggestion
Now try retrieving the availability by calling the `booking.GetAvailability` endpoint through the API Explorer in the Development Dashboard.
```

> 🥐 Somewhere inside the `booking` package, add the following functions and import the `slices` package:

Reviewer: ```suggestion
🥐 Add the following functions inside the `booking` package, and import the `slices` package:
```

```go
func listBookingsBetween(
	ctx context.Context,
	start, end time.Time,
) ([]*Booking, error) {
	rows, err := query.ListBookingsBetween(ctx, db.ListBookingsBetweenParams{
		StartTime: pgtype.Timestamp{Time: start, Valid: true},
		EndTime:   pgtype.Timestamp{Time: end, Valid: true},
	})
	if err != nil {
		return nil, err
	}
	var bookings []*Booking
	for _, row := range rows {
		bookings = append(bookings, &Booking{
			ID:    row.ID,
			Start: row.StartTime.Time,
			End:   row.EndTime.Time,
			Email: row.Email,
		})
	}
	return bookings, nil
}

func filterBookableSlots(
	slots []BookableSlot,
	now time.Time, 
	bookings []*Booking,
) []BookableSlot {
	// Remove slots for which the start time has already passed.
	slots = slices.DeleteFunc(slots, func(s BookableSlot) bool {
		// Has the slot already passed?
		if s.Start.Before(now) {
			return true
		}

		// Is there a booking that overlaps with this slot?
		for _, b := range bookings {
			if b.Start.Before(s.End) && b.End.After(s.Start) {
				return true
			}
		}

		return false
	})
	return slots
}
```

> We will use these functions to figure out which slots are bookable and which are not to avoid double bookings. 

Reviewer: ```suggestion
We'll use these functions to figure out which slots are bookable, and which are not, to avoid double bookings. 
```

			return nil, err
		}
		slots = append(slots, daySlots...)
	}

	// Get bookings for the next 7 days.
	activeBookings, err := listBookingsBetween(ctx, fromDate, fromDate.AddDate(0, 0, numDays))
	if err != nil {
		return nil, err
	}

	slots = filterBookableSlots(slots, time.Now(), activeBookings)
	return &SlotsResponse{Slots: slots}, nil
}

func bookableSlotsForDay(date time.Time, avail *Availability) ([]BookableSlot, error) {
	if avail.Start == nil || avail.End == nil {
		return nil, nil
	}
	availStartTime, err1 := strToTime(avail.Start)
	availEndTime, err2 := strToTime(avail.End)
	if err := errors.Join(err1, err2); err != nil {
		return nil, err
	}

	availStart := date.Add(time.Duration(availStartTime.Microseconds) * time.Microsecond)
	availEnd := date.Add(time.Duration(availEndTime.Microseconds) * time.Microsecond)

	// Compute the bookable slots in this day, based on availability.
	var slots []BookableSlot
	start := availStart
	for {
		end := start.Add(DefaultBookingDuration)
		if end.After(availEnd) {
			break
		}
		slots = append(slots, BookableSlot{
			Start: start,
			End:   end,
		})
		start = end
	}

	return slots, nil
}
```

## 6. Managing scheduled bookings

> In order to display the scheduled bookings in the admin dashboard we need to add the functionality to list all bookings. While we're at it, we will also make it possible to delete bookings.

Reviewer: ```suggestion
To display the scheduled bookings in the admin dashboard, we need to add the functionality to list all bookings. While we're at it, we'll also make it possible to delete bookings.
```

🥐 Add two new endpoints to `booking/booking.go`:

```go
-- booking/booking.go --
type ListBookingsResponse struct {
	Booking []*Booking `json:"bookings"`
}

//encore:api auth method=GET path=/booking
func ListBookings(ctx context.Context) (*ListBookingsResponse, error) {
	rows, err := query.ListBookings(ctx)
	if err != nil {
		return nil, err
	}

	var bookings []*Booking
	for _, row := range rows {
		bookings = append(bookings, &Booking{
			ID:    row.ID,
			Start: row.StartTime.Time,
			End:   row.EndTime.Time,
			Email: row.Email,
		})
	}
	return &ListBookingsResponse{Booking: bookings}, nil
}

//encore:api auth method=DELETE path=/booking/:id
func DeleteBooking(ctx context.Context, id int64) error {
	return query.DeleteBooking(ctx, id)
}
```

> And that is it! We now have all the backend endpoints in place to be able to supply the frontend with data 🎉

Reviewer: ```suggestion
That's it! We now have all the backend endpoints in place to be able to supply the frontend with data. 🎉
```

## 7. Running the React frontend

The frontend should now be working as expected.

🥐 Go to [http://localhost:4000/frontend/](http://localhost:4000/frontend/) and try out your new booking system.

> The frontend is built using [React](https://react.dev/) and [tailwindcss](https://tailwindcss.com/) and make makes use of Encores ability to generate type-safe [request clients](https://encore.dev/docs/develop/client-generation). This means that you do not need to manually keep the request/response objects in sync on the frontend. To generate a client run:

Reviewer: ```suggestion
The frontend is built using [React](https://react.dev/) and [Tailwind CSS](https://tailwindcss.com/). It uses Encore's ability to generate type-safe [request clients](https://encore.dev/docs/develop/client-generation). This means you don't need to manually keep the request/response objects in sync on the frontend. To generate a client:
```

```bash
$ encore gen client <APP_NAME> --output=./src/client.ts --env=<ENV_NAME>
```

> While you are developing you are going to want to run this command quite often (whenever you make a change to your endpoints) so having it as an `npm` script is a good idea. Take a look at the scripts in the `package.json` file:

Reviewer: ```suggestion
While you're developing, you are going to want to run this command quite often (whenever you make a change to your endpoints) so having it as an `npm` script is a good idea. Take a look at the scripts in the `package.json` file:
```

```json
{
...
"scripts": {
    ...
    "gen": "encore gen client <Encore App ID> --output=./src/lib/client.ts --env=staging",
    "gen:local": "encore gen client <Encore App ID> --output=./src/lib/client.ts --env=local"
  },
}
```

For this frontend we use the request client together with [TanStack Query](https://tanstack.com/query/latest). When building something a bit more complex, you will likely need to deal with caching, refetching, and data going stale. [TanStack Query](https://tanstack.com/query/latest) is a popular library that was built to solve exactly these problems and works great with the Encore request client. 

> See our the docs page about [integrating with a web frontend](docs/how-to/integrate-frontend) to learn more about these subjects. 

Reviewer: ```suggestion
See our the docs page about [integrating with a web frontend](docs/how-to/integrate-frontend) to learn more. 
```

```shell
$ git add -A .
$ git commit -m 'Initial commit'
$ git push encore
```

Encore will now build and test your app, provision the needed infrastructure, and deploy your application to the cloud.

After triggering the deployment, you will see a URL where you can view its progress in Encore's [Cloud Dashboard](https://app.encore.dev). It will look something like: `https://app.encore.dev/$APP_ID/deploys/...`

From there you can also see metrics, traces, link your app to a GitHub repo to get automatic deploys on new commits, and connect your own AWS or GCP account to use for production deployment.

🥐 When the deploy has finished, you can try out your booking system by going to `https://staging-$APP_ID.encr.app/frontend/`.

*You now have an Appointment Booking System running in the cloud, well done!*

## 8. Sending confirmation emails using SendGrid

In order for the users to get a confirmation email when they book an appointment we need to add an email integration. 

Conveniently for us, there is a ready to use SendGrid integration as an [Encore Bit](https://github.com/encoredev/examples?tab=readme-ov-file#bits). 

> 🥐 Follow the instruction https://github.com/encoredev/examples/tree/main/bits/sendgrid to add the SendGrid integration to your project.

Reviewer: ```suggestion
🥐 [Follow the instructions](https://github.com/encoredev/examples/tree/main/bits/sendgrid) to add the SendGrid integration to your project.
```
